ckanext/validation/utils.py

```diff
@@ -1,73 +1,304 @@
+    stream = BytesIO(file.read())
+    file.seek(0, os.SEEK_END)
+
+    return stream
+
+
+def run_async_validation(resource_id):
+    context = {u'ignore_auth': True}
+    data_dict = {u'resource_id': resource_id, u'async': True}
 
     try:
-        if os.path.exists(path):
-            os.remove(path)
-
-        first_directory = os.path.split(path)[0]
-        if first_directory.endswith(u'resources'):
-            return
-        if not os.listdir(first_directory):
-            os.rmdir(first_directory)
-
-        second_directory = os.path.split(first_directory)[0]
-        if second_directory.endswith(u'resources'):
-            return
-        if not os.listdir(second_directory):
-            os.rmdir(second_directory)
-
-    except OSError as e:
-        log.warning(u'Error deleting uploaded file: %s', e)
+        tk.get_action(u'resource_validation_run')(context, data_dict)
+    except tk.ValidationError as e:
+        log.warning(u'Could not run validation for resource {}: {}'.format(
+            resource_id, e))
+
+
+def _is_resource_requires_validation(context, data_dict):
+    """Check if new resource requires validation"""
+    schema = data_dict.get(u'schema')
+
+    if asbool(data_dict.get('align_default_schema')):
+        schema = _get_default_schema(data_dict["package_id"])
+
+    if not schema:
+        log.info("Missing schema. Skipping validation")
+        return False
+
+    if not data_dict.get(u'format'):
+        return False
+
+    for plugin in plugins.PluginImplementations(IDataValidation):
+        if not plugin.can_validate(context, data_dict):
+            log.debug('Skipping validation for new resource')
+            return False
+
+    res_format = data_dict.get(u'format', u'').lower()
+    supportable_format = res_format in settings.SUPPORTED_FORMATS
+
+    if supportable_format and (data_dict.get(u'url_type') == u'upload'
+                               or data_dict.get(u'url')):
+        return True
+
+    return False
+
+
+def _get_default_schema(package_id):
+    """Dataset could have a default_schema, that could be used
+    to validate resource"""
+    dataset = tk.get_action(u'package_show')({
+        'ignore_auth': True
+    }, {
+        'id': package_id
+    })
+
+    return dataset.get(u'default_data_schema')
+
+
+def _is_updated_resource_requires_validation(context, old_resource,
+                                             new_resource):
+    res_id = new_resource["id"]
+    is_creation = not old_resource
+
+    schema = new_resource.get(u'schema')
+
+    if not schema:
+        log.info("Missing resource schema. Skipping validation")
+        return False
+
+    if not new_resource.get(u'format'):
+        return False
+
+    for plugin in plugins.PluginImplementations(IDataValidation):
+        if not plugin.can_validate(context, new_resource):
+            log.debug('Skipping validation for resource {}'.format(res_id))
+            return False
+
+    if new_resource.get(u'upload'):
```

Comment:
These three conditions could be combined.

Reply:
I prefer to leave it as is, because the only point of doing it is to save a few bytes.